permission/backend.go

```diff
@@ -0,0 +1,280 @@
+	node               *node.Node
+	ethClnt            bind.ContractBackend
+	eth                *eth.Ethereum
+	key                *ecdsa.PrivateKey
+	dataDir            string
+	permConfig         *types.PermissionConfig
+	contract           ptype.InitService
+	backend            ptype.Backend
+	useDns             bool
+	isRaft             bool
+	startWaitGroup     *sync.WaitGroup // waitgroup to make sure all dependencies are ready before we start the service
+	errorChan          chan error      // channel to capture error when starting aysnc
+	networkInitialized bool
+	controlService     ptype.ControlService
+}
+
+const (
+	PERMISSION_EEA   = "EEA"
+	PERMISSION_BASIC = "BASIC"
+)
+
+var permissionService *PermissionCtrl
+
+func setPermissionService(ps *PermissionCtrl) {
+	if permissionService == nil {
+		permissionService = ps
+	}
+}
+
+// Create a service instance for permissioning
+//
+// Permission Service depends on the following:
+// 1. EthService to be ready
+// 2. Downloader to sync up blocks
+// 3. InProc RPC server to be ready
+func NewQuorumPermissionCtrl(stack *node.Node, pconfig *types.PermissionConfig, useDns bool) (*PermissionCtrl, error) {
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+
+	if pconfig.PermissionsModel == "" {
+		pconfig.PermissionsModel = PERMISSION_BASIC
+	}
+	p := &PermissionCtrl{
+		node:           stack,
+		key:            stack.GetNodeKey(),
+		dataDir:        stack.DataDir(),
+		permConfig:     pconfig,
+		startWaitGroup: wg,
+		errorChan:      make(chan error),
+		useDns:         useDns,
+		isRaft:         false,
+	}
+
+	p.populateBackEnd()
+	stopChan, stopSubscription := ptype.SubscribeStopEvent()
+	inProcRPCServerSub := stack.EventMux().Subscribe(rpc.InProcServerReadyEvent{})
+	log.Debug("permission service: waiting for InProcRPC Server")
+
+	go func(_wg *sync.WaitGroup) {
+		defer func(start time.Time) {
+			log.Debug("permission service: InProcRPC server is ready", "took", time.Since(start))
+			stopSubscription.Unsubscribe()
+			inProcRPCServerSub.Unsubscribe()
+			_wg.Done()
+		}(time.Now())
+		select {
+		case <-inProcRPCServerSub.Chan():
+		case <-stopChan:
+		}
+	}(wg) // wait for inproc RPC to be ready
+	return p, nil
+}
+
+func (p *PermissionCtrl) Start(srvr *p2p.Server) error {
+	log.Info("permission service: starting")
+	go func() {
+		log.Info("permission service: starting async")
+		p.asyncStart()
+	}()
+	return nil
+}
+
+func (p *PermissionCtrl) Protocols() []p2p.Protocol {
+	return []p2p.Protocol{}
+}
+
+func (p *PermissionCtrl) APIs() []rpc.API {
+	return []rpc.API{
+		{
+			Namespace: "quorumPermission",
+			Version:   "1.0",
+			Service:   NewQuorumControlsAPI(p),
+			Public:    true,
+		},
+	}
+}
+
+func (p *PermissionCtrl) Stop() error {
+	log.Info("permission service: stopping")
+	ptype.StopFeed.Send(ptype.StopEvent{})
+	log.Info("permission service: stopped")
+	return nil
+}
+
+func (p *PermissionCtrl) IsEEAPermission() bool {
+	return p.permConfig.PermissionsModel == PERMISSION_EEA
```

Comment:
I think it's worthwhile making this case insensitive and/or adding a check for a valid permission model.

At the moment, if someone sets `permissionModel` to "eea" then quorum will actually use the basic permission model and no warning will be given. Similarly, if the user has a typo and sets it to (say) "EEEA" then the basic permission model would get used and no warning given.

Reply:
done